### DISCUSSION TITLE
Gdk-ERROR using fltk-webview

### DISCUSSION THREAD

**Discussion by interstellarvoyager** [fltk-rs/fltk-rs#1272] [Q&A] [answered]
in my Cargo.toml i have :
```
[dependencies]                    
fltk = "^1.3"
fltk-webview = "^0.2"
```
in main.rs i have : 
```
	let w = 2*600;
	let h = 2*400;

	let app = app::App::default();
	let mut win = DoubleWindow::default()
		.with_size(w, h)
		.center_screen();

	let mut wv_win = Window::default()
		.with_size(w-20, h-20)
		.center_of_parent();
	wv_win.end();

	win.make_resizable(true);
	win.end();
	win.show();

	let wv = Webview::create(false, &mut wv_win);
	wv.navigate("https://www.wikipedia.org");

	app.run().unwrap();
```
and the error i get is : 
```
(WebKitWebProcess:394077): Gdk-ERROR **: 12:27:37.608: The program 'WebKitWebProcess' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadValue (integer parameter out of range for operation)'.
  (Details: serial 198 error_code 2 request_code 152 (GLX) minor_code 34)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

**Comment by MoAlyousef** [comment, marked as answer]
I don't think it's fltk-webview issue, a similar issue can be found here:
https://bugzilla.redhat.com/show_bug.cgi?id=1311519

It depends I guess on the libwekit2gtk also on the hardware acceleration on your system. 
What system are you running?
What's the output of "glxinfo -B" and "uname -a"?
Also what compositor (and desktop environment) are you using?

**Reply by interstellarvoyager** [reply]
i'm on linux mint. the output of "glxinfo -B" is 
```
name of display: :0
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  152 (GLX)
  Minor opcode of failed request:  24 (X_GLXCreateNewContext)
  Value in failed request:  0x0
  Serial number of failed request:  52
  Current serial number in output stream:  53
```
and the output of "uname -a" is 
```
Linux linuxmint-workstation 5.4.0-122-generic #138-Ubuntu SMP Wed Jun 22 15:00:31 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```
i'm pretty sure i'm using the cinnamon desktop environment, as for the compositor, i couldn't find out

**Reply by interstellarvoyager** [reply]
interestingly, if i run "$ cinnamon" i get : 
```
(cinnamon:399138): Clutter-CRITICAL **: 14:13:40.976: Unable to initialize Clutter: Unable to initialize the Clutter backend: no available drivers found.
Cinnamon error: Unable to initialize Clutter.
```

**Reply by interstellarvoyager** [reply]
i think i know what happened. i updated my nvidia driver earlier in the day and totally forgot to restart my system. perhaps that's where my troubles originated because, now that i did that, i'm no longer getting that error

**Comment by interstellarvoyager** [comment]
there is one more issue; i haven't gotten anything to display in the Webview
![Screenshot from 2022-08-09 14-50-00](https://user-images.githubusercontent.com/10271585/183748625-5c635ac9-7ef1-4d7a-ab27-77c7b46d4d4c.png)

**Reply by MoAlyousef** [reply]
What’s the output of `glxinfo -B` after restarting the system?

Unfortunately the abundance of compositors and desktop envs on Linux, all with their own peculiarities, makes embedding webview on Linux unreliable. It might work on some system and then breaks on another user’s system.

**Reply by interstellarvoyager** [reply]
the output of ```glxinfo -B``` now is : 
```
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Memory info (GL_NVX_gpu_memory_info):
    Dedicated video memory: 2048 MB
    Total available memory: 2048 MB
    Currently available dedicated video memory: 1670 MB
OpenGL vendor string: NVIDIA Corporation
OpenGL renderer string: NVIDIA GeForce GTX 1050/PCIe/SSE2
OpenGL core profile version string: 4.6.0 NVIDIA 515.65.01
OpenGL core profile shading language version string: 4.60 NVIDIA
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.6.0 NVIDIA 515.65.01
OpenGL shading language version string: 4.60 NVIDIA
OpenGL context flags: (none)
OpenGL profile mask: (none)

OpenGL ES profile version string: OpenGL ES 3.2 NVIDIA 515.65.01
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```
are there also other ways of embedding webview within fltk-rs ?

**Reply by MoAlyousef** [reply]
Unfortunately no ready-made solution except for fltk-webview. You can try modifying the [code here](https://github.com/fltk-rs/fltk-webview/blob/86b0d590c5127f01a8b010dc718d081d86577ace/src/lib.rs#L176) to `if !win_manager("gnome-session") {` to see if it helps. Otherwise it would be a problem in the x11 code [here](https://github.com/fltk-rs/fltk-webview/blob/86b0d590c5127f01a8b010dc718d081d86577ace/fltk-webview-sys/src/gtk_helper.c#L14).

**Reply by interstellarvoyager** [reply]
i'm down, but are there any docs or guides around for novices that could help guide the way ?

**Reply by MoAlyousef** [reply]
I'm afraid not. X11 embedding is complicated and some compositors (mostly gnome) just don't respect the XEmbed protocol.

**Reply by interstellarvoyager** [reply]
that still lights the way. i could google "X11 embedding", "compositors", and "XEmbed protocol" to get me started. i won't see to it soon. i'm just gonna switch projects for a while until i hit a road block there and then come back to this and check it out. thanks for your help, man
